src/vidpf.rs

```diff
@@ -384,22 +435,110 @@ impl From<VidpfServerId> for Choice {
-#[derive(Debug)]
+/// Vidpf correction word
+///
+///  Adjusts values of shares during the VIDPF evaluation.
+#[derive(Clone, Debug)]
 struct VidpfCorrectionWord<W: VidpfValue> {
     seed: VidpfSeed,
     left_control_bit: Choice,
     right_control_bit: Choice,
     weight: W,
 }
 
+impl<W: VidpfValue> ConstantTimeEq for VidpfCorrectionWord<W> {
+    fn ct_eq(&self, other: &Self) -> Choice {
+        self.seed.ct_eq(&other.seed)
+            & self.left_control_bit.ct_eq(&other.left_control_bit)
+            & self.right_control_bit.ct_eq(&other.right_control_bit)
+            & self.weight.ct_eq(&other.weight)
+    }
+}
+
+impl<W: VidpfValue> PartialEq for VidpfCorrectionWord<W>
+where
+    W: ConstantTimeEq,
+{
+    fn eq(&self, other: &Self) -> bool {
+        self.ct_eq(other).into()
+    }
+}
+
+impl<W: VidpfValue> Encode for VidpfCorrectionWord<W> {
+    fn encode(&self, _bytes: &mut Vec<u8>) -> Result<(), CodecError> {
+        todo!();
+    }
+
+    fn encoded_len(&self) -> Option<usize> {
+        todo!();
+    }
+}
+
+impl<W: VidpfValue> ParameterizedDecode<W::ValueParameter> for VidpfCorrectionWord<W> {
+    fn decode_with_param(
+        _decoding_parameter: &W::ValueParameter,
+        _bytes: &mut Cursor<&[u8]>,
+    ) -> Result<Self, CodecError> {
+        todo!();
+    }
+}
+
+/// Vidpf public share
+///
 /// Common public information used by aggregation servers.
-#[derive(Debug)]
+#[derive(Clone, Debug, PartialEq)]
 pub struct VidpfPublicShare<W: VidpfValue> {
     cw: Vec<VidpfCorrectionWord<W>>,
     cs: Vec<VidpfProof>,
 }
 
+impl<W: VidpfValue> Encode for VidpfPublicShare<W> {
+    fn encode(&self, bytes: &mut Vec<u8>) -> Result<(), CodecError> {
+        for word in &self.cw {
+            word.encode(bytes)?;
+        }
+        for proof in &self.cs {
+            bytes.extend_from_slice(proof);
+        }
+        Ok(())
+    }
+
+    fn encoded_len(&self) -> Option<usize> {
+        let mut total = 0;
+        for word in &self.cw {
+            total += word.encoded_len()?;
+        }
+        for _ in &self.cs {
+            total += VIDPF_PROOF_SIZE;
+        }
+        Some(total)
+    }
+}
+
+impl<W: VidpfValue> ParameterizedDecode<(usize, W::ValueParameter)> for VidpfPublicShare<W> {
+    fn decode_with_param(
+        (bits, weight_parameter): &(usize, W::ValueParameter),
+        bytes: &mut Cursor<&[u8]>,
+    ) -> Result<Self, CodecError> {
+        let mut cw = Vec::<VidpfCorrectionWord<W>>::with_capacity(*bits);
+        let mut cs = Vec::<VidpfProof>::with_capacity(*bits);
+        for _ in 0..*bits {
+            cw.push(VidpfCorrectionWord::<W>::decode_with_param(
+                weight_parameter,
+                bytes,
+            )?);
+        }
+        for _ in 0..*bits {
+            let mut proof = VidpfProof::from([0u8; VIDPF_PROOF_SIZE]);
+            bytes.read_exact(&mut proof)?;
+            cs.push(proof)
+        }
+        Ok(Self { cw, cs })
+    }
+}
```

Comment:
Let's stub these out because we're going to need to rewrite them anyway (we have to split the control bits out of each word)
```suggestion
impl<W: VidpfValue> Encode for VidpfPublicShare<W> {
    fn encode(&self, _bytes: &mut Vec<u8>) -> Result<(), CodecError> {
        todo!()
    }

    fn encoded_len(&self) -> Option<usize> {
        todo!()
    }
}

impl<W: VidpfValue> ParameterizedDecode<(usize, W::ValueParameter)> for VidpfPublicShare<W> {
    fn decode_with_param(
        (_bits, _weight_parameter): &(usize, W::ValueParameter),
        _bytes: &mut Cursor<&[u8]>,
    ) -> Result<Self, CodecError> {
        todo!()
    }
}
```